### PR TITLE
Stbas

### PR DESCRIPTION
**Pull request by mirzaees**
Add option for short temporal baseline inversion by setting off diagonal covariance matrix elements after a bandwidth to zero

### CODE REVIEW

**Comment by scottstanie** [comment]
This looks like a cool addition, and I couldn't see a simpler way to cut out the unused computations when forming the covariance matrix.

Just a question- do you want to use either EMI or EVD for this? Or would we always take the shorter option and pick EVD if we're already going to zero out a bunch of covariance entries?

File: src/dolphin/phase_link/_core.py [hunk 1]
```diff
@@ -300,6 +308,11 @@ def run_cpl(
         strides,
         neighbor_arrays=neighbor_arrays,
     )
+    if baseline_lag:
+        iu = np.triu_indices(C_arrays.shape[0], baseline_lag)
+        il = np.tril_indices(C_arrays.shape[0], -baseline_lag)
+        C_arrays = C_arrays.at[iu].set(0)
+        C_arrays = C_arrays.at[il].set(0)
```
Comment (scottstanie) [inline src/dolphin/phase_link/_core.py:315]: Ah, this is actually the wrong indexing... the `C_arrays` shape semantics are (`pixel rows, pixel cols, time1, time2`).
So right now, this is zeroing out all the pixels in the image in a tri-diagonal pattern.

```suggestion
    if baseline_lag:
        iu = np.triu_indices(C_arrays.shape[0], baseline_lag)
        il = np.tril_indices(C_arrays.shape[0], -baseline_lag)
        C_arrays = C_arrays.at[:, :, iu].set(0)
        C_arrays = C_arrays.at[:, :, il].set(0)
```


I saw that by [changing the test](https://github.com/isce-framework/dolphin/blob/main/tests/test_phase_link_core.py#L28-L29) (which I can't suggest here)

```

@pytest.mark.parametrize("baseline_lag", [None, 3])
@pytest.mark.parametrize("use_evd", [False, True])
def test_estimation(C_truth, slc_samples, use_evd, baseline_lag):
```

so let's add this option to the basic phase linking test in a parameter

File: src/dolphin/phase_link/_core.py [hunk 2]
```diff
@@ -270,6 +276,8 @@ def run_cpl(
     calc_average_coh : bool, default=False
         If requested, the average of each row of the covariance matrix is computed
         for the purposes of finding the best reference (highest coherence) date
+    baseline_lag : int, optional, default=None
+        lag for temporal baseline to do short temporal baseline inversion (STBAS)
```
Comment (scottstanie) [inline src/dolphin/phase_link/_core.py:280]:
```suggestion
    baseline_lag : int, optional, default=None
        StBAS parameter to include only nearest-N interferograms for phase linking.
        A `baseline_lag` of `n` will only include the closest `n` interferograms.
        `baseline_line` must be positive.
```

File: src/dolphin/workflows/config/_common.py [hunk 3]
```diff
@@ -104,6 +104,13 @@ class PhaseLinkingOptions(BaseModel, extra="forbid"):
         gt=0.0,
         lt=1.0,
     )
+    baseline_lag: Optional[int] = Field(
+        None,
+        description=(
+            "lag for temporal baseline to do short temporal baseline"
```
Comment (scottstanie) [inline src/dolphin/workflows/config/_common.py:110]: maybe we could also include the same longer description here?

File: src/dolphin/workflows/config/_common.py [hunk 4]
```diff
@@ -104,6 +104,13 @@ class PhaseLinkingOptions(BaseModel, extra="forbid"):
         gt=0.0,
         lt=1.0,
     )
+    baseline_lag: Optional[int] = Field(
+        None,
+        description=(
```
Comment (scottstanie) [inline src/dolphin/workflows/config/_common.py:109]:
```suggestion
        None,
        gt=0,
        description=(
```

(is this right? it has to be larger than 0 if passed?)

**Comment by mirzaees** [comment]
> This looks like a cool addition, and I couldn't see a simpler way to cut out the unused computations when forming the covariance matrix.
> 
> Just a question- do you want to use either EMI or EVD for this? Or would we always take the shorter option and pick EVD if we're already going to zero out a bunch of covariance entries?

Thanks @scottstanie 
I think it can stay optional, but EVD seems more appropriate as default

**Pull request merged**